logging: Add optional function name prefix

nordic-krch opened the pull request:
Extended logger to support optional log message prepending with function name.

Prepending with function name is optional (enabled by default). I've used following formatting:
`[00:00:24.229,644] <err> app_test.foo(): err message`

Fixes #10644
Fixes #10700 

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
logging: Add optional function name prefix

Extended logger to support optional log message prepending with
function name.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

jhedberg commented on include/logging/log_core.h:
```diff
@@ -188,6 +231,7 @@ extern "C" {
 	)								    \
 	))
 
+
```
This added line seems unrelated to the PR (it's also redundant)

jhedberg commented on subsys/logging/log_output.c:
```diff
@@ -177,7 +177,9 @@ static int ids_print(struct log_msg *msg,
-	total += print_formatted(log_output, "%s: ",
+	total += print_formatted(log_output,
+				IS_ENABLED(CONFIG_LOG_FUNCTION_NAME) ?
+				"%s." :"%s: ",
```
Just a minor whitespace fix to make this more readable: add a space between the : and the opening " of the second string.

jhedberg reviewed (approved):
Looks generally good, just a few coding style comments

nordic-krch commented on include/logging/log_core.h:
```diff
@@ -125,6 +148,26 @@ extern "C" {
+ *	  used.
+ */
+#if CONFIG_LOG_FUNCTION_NAME
+#define _LOG_STR(...) "%s(): " __LOG_ARG_1(__VA_ARGS__), __func__\
```
i was thinking that () will be self explained but if you think that it's not worth spending 2 characters i will remove it.

pizi-nordic replied:
I would keep keep the parentheses.

jhedberg replied:
@pizi-nordic I don't have a particularly strong opinion either way, just the general desire to keep the messages as compact as possible without sacrificing readability.

nordic-krch replied:
@jhedberg i've noticed that #10711 @jukkar added %s() in case CONFIG_LOG_FUNCTION_NAME. I'm also leaning towards keeping parentheses (self-explaining as the argument).

pfalcon replied:
>  I'm also leaning towards keeping parentheses (self-explaining as the argument).

IMHO, drop it and save bytes. Likewise, intend to be propose to drop xml-like syntax for `<err>` vs `ERR`. (On critical path comparing that to the old log output.)

nordic-krch replied:
Seems no approach to suit them all. However, i'm more convinced to change <err> to ERR that removing brackets. There is also instance level filtering. In that case messages are prefixed with <module_name>.<instance_name>.<function_name> (e.g. `foo.my_foo.init`) with those 3 words together brackets improve readability otherwise function name may be confused with instance name. Unless we use something else than `.` as the separator in one of those cases. but it has to be something that is not used in naming (`~`, `#` ?).

Some examples
- `foo.my_foo#init` and  `foo#init`
- `foo~my_foo.init` and `foo.init`

No strong opinion but imo any of those examples is less readable than `foo.my_foo.init()`.

jukkar replied:
> No strong opinion but imo any of those examples is less readable than `foo.my_foo.init()`

Using dots to separate the fields looks less crowded and easier to read than the other options.

nordic-krch commented:
it seems that because i added one additional parameter i need to extend logger to support 10 arguments and it is not straight forward since 10 arguments don't fit in 2 chunks of log message. Need to extend that.

jukkar commented:
> Once this PR from @jukkar is merged there should still only be 9 arguments required (unless I'm overlooking something non-networking related): #10711

I think #10711 should solve this argument count issue in networking side. Someone just needs to approve that change, after that I can merge it.

jukkar commented:
#10711 is now merged, @nordic-krch you could now rebase against master and re-submit

dbkinder reviewed (approved):
+1 for doc change

nordic-krch committed:
logging: Added support for 10 arguments in log message

Extended supported number of arguments in log message. Support for
messages consisting of more than 2 chunks had to be added. So far
messages could consist of one chunk (up to 3 args) or two chunks
(2 args in first chunk and 7 in second chunk). Once 2+ chunks
support is added number of arguments is techinically limited to
15 (4 bit field). log_core and log_output extended to suppor 10
arguments.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
Rebased on #10794 which adds support to 10 arguments in log api.
